base/linalg/cholesky.jl

```diff
@@ -100,14 +103,23 @@ function cholfact!{T<:BlasFloat,S,UpLo}(C::Cholesky{T,S,UpLo})
     C
 end
 
-cholfact{T<:BlasFloat}(A::StridedMatrix{T}, uplo::Symbol=:U; pivot=false, tol=0.0) = cholfact!(copy(A), uplo, pivot=pivot, tol=tol)
-function cholfact{T}(A::StridedMatrix{T}, uplo::Symbol=:U; pivot=false, tol=0.0)
-    S = promote_type(typeof(chol(one(T))),Float32)
-    S <: BlasFloat && return cholfact!(convert(AbstractMatrix{S}, A), uplo, pivot = pivot, tol = tol)
-    pivot && throw(ArgumentError("pivot only supported for Float32, Float64, Complex{Float32} and Complex{Float64}"))
-    S != T && return cholfact!(convert(AbstractMatrix{S}, A), uplo)
-    return cholfact!(copy(A), uplo)
-end
+cholfact{T<:BlasFloat}(A::StridedMatrix{T}, uplo::Symbol=:U; pivot::PivotUnion=Pivot{false}, tol=0.0) =
+    cholfact!(copy(A), uplo, pivot=pivot, tol=tol)
+
+
+copy_oftype{T}(A::StridedMatrix{T}, ::Type{T}) = copy(A)
+copy_oftype{T,S}(A::StridedMatrix{T}, ::Type{S}) = convert(AbstractMatrix{S}, A)
+cholfact{T}(A::StridedMatrix{T}, uplo::Symbol=:U; pivot::PivotUnion=Pivot{false}, tol=0.0) =
+    _cholfact(copy_oftype(A, promote_type(typeof(chol(one(T))),Float32)), pivot, uplo, tol=tol)
+
+_cholfact{T<:BlasFloat}(A::StridedMatrix{T}, pivot::PivotUnion, uplo::Symbol=:U; tol=0.0) =
+    cholfact!(A, uplo, pivot = pivot, tol = tol)
```

Comment:
It seems this needs to be split into two identical methods for the different `Pivot` arguments to avoid a method ambiguity warning:

```julia
_cholfact{T<:BlasFloat}(A::StridedMatrix{T}, pivot::Type{Pivot{true}}, uplo::Symbol=:U; tol=0.0) =
    cholfact!(A, uplo, pivot = pivot, tol = tol)
_cholfact{T<:BlasFloat}(A::StridedMatrix{T}, pivot::Type{Pivot{false}}, uplo::Symbol=:U; tol=0.0) =
    cholfact!(A, uplo, pivot = pivot, tol = tol)
```